1750: Endpoint specification for verein360 (#1773)
digitalfabrik/entitlementcard: docs/verein360-enpoint.md

# Verein360 Endpoint Documentation

## Introduction
Verein360/BLSV should be able to send already approved entitlementcard application for their trainers.
An api token that can be generated from the "Digitale Druckerei" account is required to access the endpoint.
All details about the endpoint can be found in the [API Documentation](../specs/backend-api.graphql).

## Usage

> ### Set Authourization Header

Reviewer: ```suggestion
### Authorization Header
```

> Set the HTTP header `Authorization Bearer <api-token>` to access the endpoint.

Reviewer: ```suggestion
Set the HTTP `Authorization` header to `Bearer <api-token>` to access the endpoint.
```

### Call GraphQL Mutation to add an application

```graphql
mutation addEakApplication($regionId: Int!, $application: ApplicationInput!, $project: String!) {
  result: addEakApplication(regionId: $regionId, application: $application, project: $project)
}
```

### Sample Mutation Variables

```json
  {
  "regionId": 2,                                      // Must be the correct region ID where the card owner lives
  "application": {
    "applicationDetails": {
      "applicationType": "FIRST_APPLICATION",         // Must be FIRST_APPLICATION
      "cardType": "BLUE",                             // Must be BLUE
      "givenInformationIsCorrectAndComplete": true,   // Must be true
      "hasAcceptedEmailUsage": true,                  // true or false
      "hasAcceptedPrivacyPolicy": true,               // Must be true
      "wantsDigitalCard": true,                       // Must be true
      "wantsPhysicalCard": false,                     // Must be false
      "blueCardEntitlement": {
        "entitlementType": "WORK_AT_ORGANIZATIONS",   // Must be WORK_AT_ORGANIZATIONS
        "workAtOrganizationsEntitlement": {
          "list": {

>             "amountOfWork": 7.5,

Reviewer: :wrench: Should we add info on the unit? Hours/Week? Hours/Month?

            "organization": {
              "address": {
                "country": {

>                   "shortText": "Germany"

Reviewer: :question: Why are these `shortText` attributes needed?
Not sure about this one:
```suggestion
                  "shortText": "Deutschland"
```

Reviewer 2: if you send an application you can check the actual payload. But good question why this was implemented like this, maybe @michael-markl can give some input

                },
                "houseNumber": {
                  "shortText": "123"
                },
                "location": {

>                   "shortText": "Munich"

Reviewer: :wrench: Same for occurrences below:
```suggestion
                  "shortText": "München"
```

                },
                "postalCode": {

>                   "shortText": "80331"

Reviewer: :thinking: 
```suggestion
                  "shortText": 80331
```

Reviewer 2: its a string. thats okay

                },
                "street": {
                  "shortText": "Example Street"
                }
              },
              "category": {"shortText": "Sport"},     // Must be Sport
              "contact": {
                "email": {"email": "jane.doe@sportverein.de"},
                "hasGivenPermission": true,           // Must be true
                "name": {"shortText": "Jane Doe"},
                "telephone": {"shortText": "0150123456789"}
              },
              "name": {
                "shortText": "Sportverein Augsburg-Nord"
              }
            },
            "payment": false,                         // Must be false
            "responsibility": {"shortText": "Trainer"},
            "workSinceDate": {"date": "2020-10-06"},

>             "isAlreadyVerified": true                 // Must be true, so that we can mark the application as verified. 

Author: Please let me know what we want to do in this case.
We could
1) Send no error message and just accept the applicaiton and not mark it as verified
2) Send an unauthorized error message and not save the application

I think I would prefer 2) to ensure verein360 fixes auth problems and not make the municipalities a lot of work of verifying applications manullay.

Reviewer 2: But as we discussed this source may also be defined via the purpose of the bearer token or a particular role for verein 360. We may add some meta data for API Tokens in general

                                                      // If this is set to true, but not valid application token is set 
                                                      // the application is rejected and an error code returned.
          }
        }
      }
    },
    "personalData": {
      "address": {
        "country": {
          "shortText": "Germany"
        },
        "houseNumber": {
          "shortText": "123"
        },
        "location": {
          "shortText": "Munich"
        },
        "postalCode": {
          "shortText": "80331"
        },
        "street": {
          "shortText": "Example Street"
        }
      },
      "dateOfBirth": {
        "date": "1990-01-01"
      },
      "emailAddress": {
        "email": "johndoe@example.com"
      },
      "forenames": {
        "shortText": "John"
      },
      "surname": {
        "shortText": "Doe"
      },
      "telephone": {
        "shortText": "123456789"
      }
    }
  },
  "project": "bayern.ehrenamtskarte.app"              // Must be bayern.ehrenamtskarte.app  
}
```

### Get Region ID for application

#### Get all Regions

> The region ID can be obtained by calling the following query:

Reviewer: ```suggestion
All available regions and their IDs can be obtained by calling the following graphql endpoint:
```

## Usage

> ### Authourization Header

Reviewer: :upside_down_face: Typo
```suggestion
### Authorization Header
```
